src/Lib.P2P/SecureCommunication/Psk1Stream.cs

```diff
@@ -0,0 +1,203 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+using Lib.P2P.Cryptography;
+using Org.BouncyCastle.Crypto;
+using Org.BouncyCastle.Crypto.Engines;
+using Org.BouncyCastle.Crypto.Parameters;
+
+namespace Lib.P2P.SecureCommunication
+{
+    /// <summary>
+    ///   A duplex stream that is encrypted with a <see cref="PreSharedKey"/>.
+    /// </summary>
+    /// <remarks>
+    ///   The XSalsa20 cipher is used to encrypt the data.
+    /// </remarks>
+    /// <seealso href="https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md"/>
+    public class Psk1Stream : Stream
+    {
+        private const int KeyBitLength = 256;
+        private const int NonceBitLength = 192;
+        private const int NonceByteLength = NonceBitLength / 8;
+
+        private Stream _stream;
+        private PreSharedKey _key;
+        private IStreamCipher _readCipher;
+        private IStreamCipher _writeCipher;
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="Psk1Stream"/> class. 
+        /// </summary>
+        /// <param name="stream">
+        ///   The source/destination of the unprotected stream.
+        /// </param>
+        /// <param name="key">
+        ///   The pre-shared 256-bit key for the private network of peers.
+        /// </param>
+        public Psk1Stream(Stream stream,
+            PreSharedKey key)
+        {
+            if (key.Length != KeyBitLength)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)